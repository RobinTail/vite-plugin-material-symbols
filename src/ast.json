{
  "type": "Program",
  "start": 0,
  "end": 604,
  "body": [
    {
      "type": "VariableDeclaration",
      "start": 189,
      "end": 419,
      "kind": "const",
      "declarations": [
        {
          "type": "VariableDeclarator",
          "start": 195,
          "end": 418,
          "id": {
            "type": "Identifier",
            "start": 195,
            "end": 199,
            "name": "Test"
          },
          "init": {
            "type": "ArrowFunctionExpression",
            "start": 202,
            "end": 418,
            "async": false,
            "expression": true,
            "generator": false,
            "params": [],
            "body": {
              "type": "CallExpression",
              "start": 224,
              "end": 418,
              "optional": false,
              "_rollupAnnotations": [
                { "end": 223, "start": 208, "type": "pure" }
              ],
              "callee": {
                "type": "Identifier",
                "start": 224,
                "end": 228,
                "name": "jsxs"
              },
              "arguments": [
                {
                  "type": "Literal",
                  "start": 229,
                  "end": 234,
                  "value": "div",
                  "raw": "\"div\""
                },
                {
                  "type": "ObjectExpression",
                  "start": 236,
                  "end": 417,
                  "properties": [
                    {
                      "type": "Property",
                      "start": 238,
                      "end": 415,
                      "method": false,
                      "shorthand": false,
                      "computed": false,
                      "key": {
                        "type": "Identifier",
                        "start": 238,
                        "end": 246,
                        "name": "children"
                      },
                      "value": {
                        "type": "ArrayExpression",
                        "start": 248,
                        "end": 415,
                        "elements": [
                          {
                            "type": "CallExpression",
                            "start": 268,
                            "end": 299,
                            "optional": false,
                            "_rollupAnnotations": [
                              { "end": 267, "start": 252, "type": "pure" }
                            ],
                            "callee": {
                              "type": "Identifier",
                              "start": 268,
                              "end": 271,
                              "name": "jsx"
                            },
                            "arguments": [
                              {
                                "type": "Identifier",
                                "start": 272,
                                "end": 276,
                                "name": "Icon"
                              },
                              {
                                "type": "ObjectExpression",
                                "start": 278,
                                "end": 298,
                                "properties": [
                                  {
                                    "type": "Property",
                                    "start": 280,
                                    "end": 296,
                                    "method": false,
                                    "shorthand": false,
                                    "computed": false,
                                    "key": {
                                      "type": "Identifier",
                                      "start": 280,
                                      "end": 288,
                                      "name": "children"
                                    },
                                    "value": {
                                      "type": "Literal",
                                      "start": 290,
                                      "end": 296,
                                      "value": "home",
                                      "raw": "\"home\""
                                    },
                                    "kind": "init"
                                  }
                                ]
                              }
                            ]
                          },
                          {
                            "type": "CallExpression",
                            "start": 319,
                            "end": 359,
                            "optional": false,
                            "_rollupAnnotations": [
                              { "end": 318, "start": 303, "type": "pure" }
                            ],
                            "callee": {
                              "type": "Identifier",
                              "start": 319,
                              "end": 322,
                              "name": "jsx"
                            },
                            "arguments": [
                              {
                                "type": "Identifier",
                                "start": 323,
                                "end": 327,
                                "name": "Icon"
                              },
                              {
                                "type": "ObjectExpression",
                                "start": 329,
                                "end": 358,
                                "properties": [
                                  {
                                    "type": "Property",
                                    "start": 331,
                                    "end": 356,
                                    "method": false,
                                    "shorthand": false,
                                    "computed": false,
                                    "key": {
                                      "type": "Identifier",
                                      "start": 331,
                                      "end": 339,
                                      "name": "children"
                                    },
                                    "value": {
                                      "type": "Literal",
                                      "start": 341,
                                      "end": 356,
                                      "value": "chevron_right",
                                      "raw": "\"chevron_right\""
                                    },
                                    "kind": "init"
                                  }
                                ]
                              }
                            ]
                          },
                          {
                            "type": "CallExpression",
                            "start": 379,
                            "end": 413,
                            "optional": false,
                            "_rollupAnnotations": [
                              { "end": 378, "start": 363, "type": "pure" }
                            ],
                            "callee": {
                              "type": "Identifier",
                              "start": 379,
                              "end": 382,
                              "name": "jsx"
                            },
                            "arguments": [
                              {
                                "type": "Identifier",
                                "start": 383,
                                "end": 387,
                                "name": "Icon"
                              },
                              {
                                "type": "ObjectExpression",
                                "start": 389,
                                "end": 412,
                                "properties": [
                                  {
                                    "type": "Property",
                                    "start": 391,
                                    "end": 410,
                                    "method": false,
                                    "shorthand": false,
                                    "computed": false,
                                    "key": {
                                      "type": "Identifier",
                                      "start": 391,
                                      "end": 399,
                                      "name": "children"
                                    },
                                    "value": {
                                      "type": "Literal",
                                      "start": 401,
                                      "end": 410,
                                      "value": "comment",
                                      "raw": "\"comment\""
                                    },
                                    "kind": "init"
                                  }
                                ]
                              }
                            ]
                          }
                        ]
                      },
                      "kind": "init"
                    }
                  ]
                }
              ]
            },
            "id": null
          }
        }
      ]
    }
  ],
  "sourceType": "module"
}
