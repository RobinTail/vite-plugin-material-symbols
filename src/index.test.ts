import { describe, expect, it, mock } from "bun:test";
import { fail } from "node:assert/strict";
import type { ModuleInfo, PluginContext } from "rollup";
import ast from "../tools/ast.json";
import plugin from "./index";

describe("Entrypoint", () => {
  it("should be a function", () => {
    expect(plugin).toBeFunction();
  });

  it("should return object of certain structure", () => {
    expect(plugin()).toEqual({
      enforce: "pre",
      moduleParsed: expect.any(Function),
      name: "material-symbols",
      transformIndexHtml: expect.any(Function),
    });
  });

  describe.each([false, true])("Operation %#", (preload) => {
    const { moduleParsed, transformIndexHtml } = plugin({ preload });
    if (!moduleParsed) fail("no moduleParsed hook");
    if (typeof moduleParsed !== "function")
      fail("moduleParsed is not a function");
    if (!transformIndexHtml) fail("no transformIndexHtml");
    if (typeof transformIndexHtml !== "function")
      fail("transformIndexHtml is not a function");

    it("injects the link to all symbols in dev mode", () => {
      const result = transformIndexHtml(
        `<html lang="en"><head><title>test</title></head></html>`,
        { path: ".", filename: "index.html" },
      );
      expect(result).toMatchSnapshot();
    });

    it("should find icon names", () => {
      const debug = mock();
      moduleParsed.call(
        { debug } as unknown as PluginContext,
        { ast, id: "file.tsx" } as unknown as ModuleInfo,
      );
      expect(debug).toHaveBeenCalledTimes(3);
      expect(debug.mock.calls).toEqual([
        [{ id: "file.tsx", message: "home" }],
        [{ id: "file.tsx", message: "chevron_right" }],
        [{ id: "file.tsx", message: "comment" }],
      ]);
    });

    it("injects the link with found icon names into html", () => {
      const result = transformIndexHtml(
        `<html lang="en"><head><title>test</title></head></html>`,
        { path: ".", filename: "index.html" },
      );
      expect(result).toMatchSnapshot();
    });
  });
});

describe("System", () => {
  it("injects the link into index.html", async () => {
    // @see https://github.com/oven-sh/bun/issues/3768
    await Bun.$`NODE_ENV=production bun --bun vite -c tools/vite.config.ts build`;
    const result = await Bun.file("./dist/tools/index.html").text();
    expect(
      result.includes(
        `<link rel="stylesheet" ` +
          `href="https://fonts.googleapis.com/css2?` +
          "family=Material+Symbols+Outlined:opsz,wght,FILL,GRAD@24,400,0,0&amp;" +
          `icon_names=chevron_right,comment,home">`,
      ),
    ).toBeTrue();
  });
});
