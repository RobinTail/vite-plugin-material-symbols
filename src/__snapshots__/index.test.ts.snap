// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Entrypoint Operation 0 injects the link to all symbols in dev mode 1`] = `
{
  "html": "<html lang="en"><head><title>test</title></head></html>",
  "tags": [
    {
      "attrs": {
        "href": "https://fonts.googleapis.com/css2?family=Material+Symbols+Outlined:opsz,wght,FILL,GRAD@24,400,0,0&",
        "rel": "stylesheet",
      },
      "injectTo": "head",
      "tag": "link",
    },
  ],
}
`;

exports[`Entrypoint Operation 0 injects the link with found icon names into html 1`] = `
{
  "html": "<html lang="en"><head><title>test</title></head></html>",
  "tags": [
    {
      "attrs": {
        "href": "https://fonts.googleapis.com/css2?family=Material+Symbols+Outlined:opsz,wght,FILL,GRAD@24,400,0,0&icon_names=chevron_right,comment,home",
        "rel": "stylesheet",
      },
      "injectTo": "head",
      "tag": "link",
    },
  ],
}
`;

exports[`Entrypoint Operation 1 injects the link to all symbols in dev mode 1`] = `
{
  "html": "<html lang="en"><head><title>test</title></head></html>",
  "tags": [
    {
      "attrs": {
        "href": "https://fonts.googleapis.com/css2?family=Material+Symbols+Outlined:opsz,wght,FILL,GRAD@24,400,0,0&",
        "rel": "stylesheet",
      },
      "injectTo": "head",
      "tag": "link",
    },
    {
      "attrs": {
        "as": "style",
        "href": "https://fonts.googleapis.com/css2?family=Material+Symbols+Outlined:opsz,wght,FILL,GRAD@24,400,0,0&",
        "rel": "preload",
      },
      "injectTo": "head-prepend",
      "tag": "link",
    },
    {
      "attrs": {
        "href": "https://fonts.gstatic.com",
        "rel": "preconnect",
      },
      "injectTo": "head-prepend",
      "tag": "link",
    },
  ],
}
`;

exports[`Entrypoint Operation 1 injects the link with found icon names into html 1`] = `
{
  "html": "<html lang="en"><head><title>test</title></head></html>",
  "tags": [
    {
      "attrs": {
        "href": "https://fonts.googleapis.com/css2?family=Material+Symbols+Outlined:opsz,wght,FILL,GRAD@24,400,0,0&icon_names=chevron_right,comment,home",
        "rel": "stylesheet",
      },
      "injectTo": "head",
      "tag": "link",
    },
    {
      "attrs": {
        "as": "style",
        "href": "https://fonts.googleapis.com/css2?family=Material+Symbols+Outlined:opsz,wght,FILL,GRAD@24,400,0,0&icon_names=chevron_right,comment,home",
        "rel": "preload",
      },
      "injectTo": "head-prepend",
      "tag": "link",
    },
    {
      "attrs": {
        "href": "https://fonts.gstatic.com",
        "rel": "preconnect",
      },
      "injectTo": "head-prepend",
      "tag": "link",
    },
  ],
}
`;
